File: ui/src/lib/types/playlist.ts
import z from "zod";
import { API } from "./api";
import { JSONBody } from "./general";
import { convertTrack, convertTrackDuplicate, Track, TrackDuplicate } from "./track";
import { convertUser, User } from "./user";

export interface Playlist {
  id: number;
  spotifyId: string;
  owner?: User;
  name: string;
  description?: string;
  public?: boolean;
  trackAmount: number;
  collaborative?: boolean;
  hasCover: boolean;
}

export interface PlaylistDuplicate extends Playlist {
  duplicates: TrackDuplicate[];
}

export interface PlaylistUnplayable extends Playlist {
  unplayables: Track[];
}

export const convertPlaylist = (playlist: API.Playlist): Playlist => {
  return {
    id: playlist.id,
    spotifyId: playlist.spotify_id,
    owner: playlist.owner ? convertUser(playlist.owner) : undefined,
    name: playlist.name,
    description: playlist.description,
    public: playlist.public,
    trackAmount: playlist.track_amount,
    collaborative: playlist.collaborative,
    hasCover: playlist.has_cover,
  }
}

export const convertPlaylists = (playlists: API.Playlist[]): Playlist[] => {
  return playlists.map(convertPlaylist)
}

export const convertPlaylistDuplicate = (p: API.PlaylistDuplicate): PlaylistDuplicate => {
  return {
    ...convertPlaylist(p),
    duplicates: p.duplicates.map(convertTrackDuplicate),
  }
}

export const convertPlaylistDuplicates = (p: API.PlaylistDuplicate[]): PlaylistDuplicate[] => {
  return p.map(convertPlaylistDuplicate)
}

export const convertPlaylistUnplayable = (p: API.PlaylistUnplayable): PlaylistUnplayable => {
  return {
    ...convertPlaylist(p),
    unplayables: p.unplayables.map(convertTrack),
  }
}

export const convertPlaylistUnplayables = (p: API.PlaylistUnplayable[]): PlaylistUnplayable[] => {
  return p.map(convertPlaylistUnplayable)
}

export const convertPlaylistsSchema = (playlists: Playlist[]): PlaylistSchema[] => {
  return playlists.map(p => ({
    id: p.id,
    name: p.name,
    trackAmount: p.trackAmount,
    hasCover: p.hasCover,
  }))
}

export const playlistSchema = z.object({
  id: z.number(),
  name: z.string(),
  trackAmount: z.number(),
  hasCover: z.boolean(),
})
export type PlaylistSchema = z.infer<typeof playlistSchema> & JSONBody;

File: internal/server/dto/playlist.go
package dto

import (
	"github.com/topvennie/sortifyr/internal/database/model"
	"github.com/topvennie/sortifyr/pkg/utils"
)

type Playlist struct {
	ID            int    `json:"id" validate:"required"`
	SpotifyID     string `json:"spotify_id"`
	Owner         User   `json:"owner,omitzero"`
	Name          string `json:"name"`
	Description   string `json:"description,omitzero"`
	Public        *bool  `json:"public,omitzero"`
	TrackAmount   int    `json:"track_amount"`
	Collaborative *bool  `json:"collaborative,omitzero"`
	HasCover      bool   `json:"has_cover"`
}

func PlaylistDTO(playlist *model.Playlist, user *model.User) Playlist {
	return Playlist{
		ID:            playlist.ID,
		SpotifyID:     playlist.SpotifyID,
		Owner:         UserDTO(user),
		Name:          playlist.Name,
		Description:   playlist.Description,
		Public:        playlist.Public,
		TrackAmount:   playlist.TrackAmount,
		Collaborative: playlist.Collaborative,
		HasCover:      playlist.CoverID != "",
	}
}

func (p Playlist) ToModel() *model.Playlist {
	return &model.Playlist{
		ID:            p.ID,
		SpotifyID:     p.SpotifyID,
		OwnerID:       p.Owner.ID,
		Name:          p.Name,
		Description:   p.Description,
		Public:        p.Public,
		TrackAmount:   p.TrackAmount,
		Collaborative: p.Collaborative,
		Owner:         *p.Owner.ToModel(),
	}
}

type PlaylistDuplicate struct {
	Playlist
	Duplicates []TrackDuplicate `json:"duplicates"`
}

func PlaylistDuplicateDTO(playlist *model.Playlist, user *model.User, duplicates []model.Track) PlaylistDuplicate {
	type trackAmount struct {
		track  model.Track
		amount int
	}

	duplicateMap := make(map[int]trackAmount)
	for i := range duplicates {
		d, ok := duplicateMap[duplicates[i].ID]
		if !ok {
			d = trackAmount{track: duplicates[i], amount: 0}
		}

		d.amount++
		duplicateMap[duplicates[i].ID] = d
	}

	return PlaylistDuplicate{
		Playlist:   PlaylistDTO(playlist, user),
		Duplicates: utils.SliceMap(utils.MapValues(duplicateMap), func(t trackAmount) TrackDuplicate { return TrackDuplicateDTO(&t.track, t.amount) }),
	}
}

type PlaylistUnplayable struct {
	Playlist
	Unplayables []Track `json:"unplayables"`
}

func PlaylistUnplayableDTO(playlist *model.Playlist, user *model.User, unplayables []model.Track) PlaylistUnplayable {
	return PlaylistUnplayable{
		Playlist:    PlaylistDTO(playlist, user),
		Unplayables: utils.SliceMap(unplayables, func(t model.Track) Track { return TrackDTO(&t) }),
	}
}
